##                          CAPSTONE_PROJECT_2_CLASSIFY_RESTAURANT_REVIEWS-NLP

**Abstract:**

This Project is a classification machine learning model which uses NLP to classify restaurant reviews(whether liked or not liked).The reviews can be anything which are related to the food of the restaurant,staff of the restaurant and also overall review of the restaurant.The Project uses 5 models namely  **i) Multinommial Naive bayes ii) Logistic Regression Model iii) Random Forest Classifier iv) Decision Tree Classifier v) Support Vector Machine(SVM) algorithm** for classifying the reviews.The classified  reviews  are  helpful  for the  restaurant to analyze  their shortcomings improve the quality of food and service in the restaurant.

**Dataset Description**

The Dataset contains 1000 reviews in text format in a ―.tsv‖ This dataset is used for training the models It is split into 75%  training data and 25% test data. The dataset contains two columns. First column contains the text reviews given by different users which are related to the food of the restaurant  as well  as the  overall review of the restaurant. The  second column contains  the sentiment i.e.  if the  review is positive  or negative.i,e 1  indicates that the review is positive and 0 indicates the review is negative. 

The dataset is imported and converted and into a Pandas Dataframe. The model we are building should predict if the review is positive or negative. The dataset is cleaned from 1000 reviews and the reviews which are not proper are discarded from the dataset and then the Dataframe is then served as input to the Count Vectorizer

### **1]Problem Statement 1:**                                       
  ***Perform Basic EDA a. Boxplot b. Histogram – Distribution of Target Variable c. Distribution Plot – Target Variable d. Aggregation for all numerical Columns e. Unique Values across all columns f. Duplicate values across all columns g. Correlation – Heatmap h. Regression Plot i. Bar Plot j. Pair plot P.s: All the above charts should be plotted against every independent variable vs Target Variable. (Except b and c)***

###  **2] Statement-2. Drop all duplicate rows**

### **3] Statement-3. Drop all non-essential feature**

### **Solution:-** 

As data has only 2 rows,one dependent and one independent we dont have numerical data for Exploratory Data Analysis
so all the above mentioned plots and comparison will not be possible, as well as no duplicate rows because chance is that texts might be same but different persion, and no NON -ESSENTIAL features

In [3]:
import pandas as pd
import numpy as np

In [4]:
df=pd.read_csv('Rest.csv')
df.head()

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1

## **4] Statement-4:: Perform basic text processing on the data.**
## **5] Staement-5: Tokenize and stem the data, so that it can be used for classification.**
## **6] Statement-6: Use NER if required**

### ***Cleaning the texts***

In [5]:
import re
import nltk                                 ## Importing natural language tool kit
nltk.download('stopwords')                  ## downloading Stop words from NLTK
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer  ## Importing stemmmer to find the root word
corp = []


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Sachy\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### ***Now Creating loop for each row to clean every word in row***

In [6]:
for i in range(0, 1000):
  review = re.sub('[^a-zA-Z]', ' ', df['Review'][i])
  review = review.lower()
  review = review.split()
  ps = PorterStemmer()
  all_stopwords = stopwords.words('english')
  all_stopwords.remove('not')                 ## We have to keep 'NOT'as it plays a vital in classyfing negative review
  review = [ps.stem(word) for word in review if not word in set(all_stopwords)]
  review = ' '.join(review)
  corp.append(review)

In [9]:
print(corp)

['wow love place', 'crust not good', 'not tasti textur nasti', 'stop late may bank holiday rick steve recommend love', 'select menu great price', 'get angri want damn pho', 'honeslti tast fresh', 'potato like rubber could tell made ahead time kept warmer', 'fri great', 'great touch', 'servic prompt', 'would not go back', 'cashier care ever say still end wayyy overpr', 'tri cape cod ravoli chicken cranberri mmmm', 'disgust pretti sure human hair', 'shock sign indic cash', 'highli recommend', 'waitress littl slow servic', 'place not worth time let alon vega', 'not like', 'burritto blah', 'food amaz', 'servic also cute', 'could care less interior beauti', 'perform', 'right red velvet cake ohhh stuff good', 'never brought salad ask', 'hole wall great mexican street taco friendli staff', 'took hour get food tabl restaur food luke warm sever run around like total overwhelm', 'worst salmon sashimi', 'also combo like burger fri beer decent deal', 'like final blow', 'found place accid could not

### ***Creating the Bag of Words model***

In [10]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
X = pd.DataFrame(cv.fit_transform(corp).toarray())
y = df.iloc[:, -1].values

In [11]:
X.head(5)

0     1     2     3     4     5     6     7     8     9     ...  1556  \
0     0     0     0     0     0     0     0     0     0     0  ...     0   
1     0     0     0     0     0     0     0     0     0     0  ...     0   
2     0     0     0     0     0     0     0     0     0     0  ...     0   
3     0     0     0     0     0     0     0     0     0     0  ...     0   
4     0     0     0     0     0     0     0     0     0     0  ...     0   

   1557  1558  1559  1560  1561  1562  1563  1564  1565  
0     0     0     0     0     0     0     0     0     0  
1     0     0     0     0     0     0     0     0     0  
2     0     0     0     0     0     0     0     0     0  
3     0     0     0     0     0     0     0     0     0  
4     0     0     0     0     0     0     0     0     0  

[5 rows x 1566 columns]

### Splitting dataset into the Training set and Test set

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

### 7. Use one-hot encoding to convert the categorical variables into numerical ones, if required. 
### 8. Build a model of choice – Classification problem statement, hence build a classification model first and calculate Confusion Matrix, AUC, F1 Score, Precision, Recall and Accuracy. 
### 9. Build at least a minimum of 4 different Regression models. All the models should use K-Fold cross Validation to train the model with at least 5-fold cross validation. 
### 10. Compare the error and pick the ideal one with least errors. 
### 11. Run hyperparameter tuning on all the models and pick the best parameters (A minimum of 2 Parameters should be tuned) and picked.

### ***Using Naive Bayes as the first Classficaion model first on the Training set as it uses conditional probability***

## ***1>> Naive Bayes Classification Model-GuassinaNB***

In [13]:
from sklearn.naive_bayes import GaussianNB,MultinomialNB
nb = GaussianNB()
mnb=MultinomialNB()
nb.fit(X_train, y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

### Predicting the Test set results

In [14]:
y_pred = nb.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [0 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [1 0]
 [1 0]
 [0 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]

### Evaluation of model using Confusion Matrix

In [15]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
cm = confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))
acc_nb=accuracy_score(y_test, y_pred)
acc_nb

              precision    recall  f1-score   support

           0       0.77      0.57      0.66       117
           1       0.69      0.85      0.76       133

    accuracy                           0.72       250
   macro avg       0.73      0.71      0.71       250
weighted avg       0.73      0.72      0.71       250



0.72

In [16]:
from sklearn.model_selection import cross_val_score
nb_Kfold_accu = cross_val_score(estimator = nb, X = X_train, y = y_train, cv = 10)
nb_Kfold_accu=nb_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(nb_Kfold_accu*100))

Accuracy: 66.67 %


### Using Multinomial Naive bayes model

In [17]:
mnb.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [18]:
y_pred = mnb.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [0 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]

### Evaluation of model using Confusion Matrix

In [19]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
cm = confusion_matrix(y_test, y_pred)
print(classification_report(y_test, y_pred))
acc_mnb=accuracy_score(y_test, y_pred)
acc_mnb
cm

              precision    recall  f1-score   support

           0       0.74      0.77      0.76       117
           1       0.79      0.77      0.78       133

    accuracy                           0.77       250
   macro avg       0.77      0.77      0.77       250
weighted avg       0.77      0.77      0.77       250



array([[ 90,  27],
       [ 31, 102]], dtype=int64)

### **K Fold Validation for model**

In [20]:
from sklearn.model_selection import cross_val_score
mnb_Kfold_accu = cross_val_score(estimator = mnb, X = X_train, y = y_train, cv = 10)
mnb_Kfold_accu=mnb_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(mnb_Kfold_accu*100))


Accuracy: 77.33 %


### Model Comparison

In [21]:
model_compare=pd.DataFrame({'Models':['GuassianNB','Multinomial'],
               'Accuracy Score':[acc_nb,acc_mnb],
            'K Fold Accuracy':[nb_Kfold_accu,mnb_Kfold_accu]})
print(model_compare)

        Models  Accuracy Score  K Fold Accuracy
0   GuassianNB           0.720         0.666667
1  Multinomial           0.768         0.773333


### ***As we observe above,Multinomial model gives better accuracy then Guassain ,we willl proceed with Multinomial model and we will try to do hyper paramter tuning for the same***

### **Hyper_Parameter Tuning**

In [22]:
from sklearn.model_selection import GridSearchCV
parameters = [{'alpha': [1,10,50,100], 'fit_prior': [True,False]}]       

In [23]:
grid_search = GridSearchCV(estimator = mnb,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
grid_search = grid_search.fit(X_train, y_train)
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.53 %
Best Parameters: {'alpha': 10, 'fit_prior': True}


In [24]:
parameters = [{'alpha': [10,11,12,13], 'fit_prior': [True]}]   
grid_search = GridSearchCV(estimator = mnb,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
grid_search = grid_search.fit(X_train, y_train)
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'alpha': 11, 'fit_prior': True}


In [25]:
parameters = [{'alpha': [10.90,10.95,11,11.05,11.1,11.15,11.20], 'fit_prior': [True]}]   
mnb_grid_search = GridSearchCV(estimator = mnb,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
mnb_grid_search = mnb_grid_search.fit(X_train, y_train)
mnb_best_accuracy = mnb_grid_search.best_score_
mnb_best_parameters = mnb_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'alpha': 11, 'fit_prior': True}


In [26]:
mnb_y_pred=mnb_grid_search.predict(X_test)
mnb_y_pred

array([0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1,
       0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1,
       1, 1, 0, 1, 1, 0, 0, 0], dtype=int64)

### **Checking Accuracy after hypertuning**

In [27]:
mnb_cm = confusion_matrix(y_test, mnb_y_pred)
print(classification_report(y_test, mnb_y_pred))
acc_mnb=accuracy_score(y_test, mnb_y_pred)
acc_mnb

              precision    recall  f1-score   support

           0       0.72      0.85      0.78       117
           1       0.85      0.71      0.78       133

    accuracy                           0.78       250
   macro avg       0.79      0.78      0.78       250
weighted avg       0.79      0.78      0.78       250



0.78

In [65]:
from sklearn.metrics import roc_auc_score
mnb_roc=roc_auc_score(y_test,mnb_y_pred)
mnb_roc

0.7844932844932846

##  ***2>> Building Logistic Regression Model***

In [28]:
from sklearn.linear_model import LogisticRegression
lg = LogisticRegression(random_state = 0)
lg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

### **K Fold Cross Validation**

In [29]:
from sklearn.model_selection import cross_val_score
lg_Kfold_accu = cross_val_score(estimator = lg, X = X_train, y = y_train, cv = 5)
lg_Kfold_accu=lg_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(lg_Kfold_accu*100))

Accuracy: 79.73 %


### **Predicting results**

In [30]:
lg_y_pred = lg.predict(X_test)
print(np.concatenate((lg_y_pred.reshape(len(lg_y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 1]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [1 1]
 [0 1]
 [1 0]
 [0 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [0 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 1]

### Evaluation of model using Confusion Matrix

In [31]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
lg_cm = confusion_matrix(y_test, lg_y_pred)
print(classification_report(y_test, lg_y_pred))
acc_lg=accuracy_score(y_test, lg_y_pred)
acc_lg


              precision    recall  f1-score   support

           0       0.73      0.84      0.78       117
           1       0.83      0.72      0.77       133

    accuracy                           0.78       250
   macro avg       0.78      0.78      0.78       250
weighted avg       0.78      0.78      0.78       250



0.776

### **Hyper Paramter Tuning**

In [32]:
parameters = [{'penalty': [11,12,'elasticnet'], 'C': [1,10,50,100,200]},
              {'tol': [0.001,0.0001,0.00001]}]
lg_grid_search = GridSearchCV(estimator = lg,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
lg_grid_search = lg_grid_search.fit(X_train, y_train)
lg_best_accuracy = lg_grid_search.best_score_
best_parameters = lg_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'tol': 0.001}


In [33]:
parameters = [{'tol': [0.01,0.001,0.0012,0.0013,0.0014]}]
lg_grid_search = GridSearchCV(estimator = lg,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
lg_grid_search = lg_grid_search.fit(X_train, y_train)
lg_best_accuracy = lg_grid_search.best_score_
best_parameters = lg_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'tol': 0.01}


In [34]:
parameters = [{'tol': [0.1,0.01,0.001]}]
lg_grid_search = GridSearchCV(estimator = lg,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
lg_grid_search = lg_grid_search.fit(X_train, y_train)
lg_best_accuracy = lg_grid_search.best_score_
best_parameters = lg_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'tol': 0.1}


In [35]:
paramters=[int(x) for x in np.linspace(start = 0.0001, stop = 0.001, num = 1)]

In [36]:
lg_grid_search = GridSearchCV(estimator = lg,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
lg_grid_search = lg_grid_search.fit(X_train, y_train)
lg_best_accuracy = lg_grid_search.best_score_
best_parameters = lg_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'tol': 0.1}


### **Prediction after Hyper Parammer tuning**

In [37]:
lg_y_pred = lg_grid_search.predict(X_test)
lg_cm = confusion_matrix(y_test, lg_y_pred)
print(classification_report(y_test, lg_y_pred))
acc_lg=accuracy_score(y_test, lg_y_pred)
acc_lg*100


              precision    recall  f1-score   support

           0       0.73      0.84      0.78       117
           1       0.83      0.72      0.77       133

    accuracy                           0.78       250
   macro avg       0.78      0.78      0.78       250
weighted avg       0.78      0.78      0.78       250



77.60000000000001

In [74]:
lg_roc=roc_auc_score(y_test,lg_y_pred)
lg_roc

0.7797056744425166

## ***3>> Building Random Forest Classifier Model***

In [38]:
from sklearn.ensemble import RandomForestClassifier
rc = RandomForestClassifier(random_state = 0)
rc.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

### **K Fold validaiton for performance of model**

In [39]:
from sklearn.model_selection import cross_val_score
rc_Kfold_accu = cross_val_score(estimator = rc, X = X_train, y = y_train, cv = 10)
rc_Kfold_accu=rc_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(rc_Kfold_accu*100))

Accuracy: 79.47 %


In [40]:
rc_y_pred = rc.predict(X_test)
print(np.concatenate((rc_y_pred.reshape(len(rc_y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [1 0]
 [0 0]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [0 1]
 [1 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 1]

### Evaluation of model using Confusion Matrix

In [41]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
cm = confusion_matrix(y_test, rc_y_pred)
print(classification_report(y_test, rc_y_pred))
rc_acc=accuracy_score(y_test, rc_y_pred)
rc_acc

              precision    recall  f1-score   support

           0       0.67      0.91      0.77       117
           1       0.88      0.61      0.72       133

    accuracy                           0.75       250
   macro avg       0.78      0.76      0.75       250
weighted avg       0.78      0.75      0.74       250



0.748

### **Hyper Tuning of Parameters**

In [42]:
params = [{'n_estimators':[100,200,300,400,500], 'criterion':['entropy','gini'],
              'max_depth':[3,4,5]}]
rc_grid_search = GridSearchCV(estimator = rc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
rc_grid_search = rc_grid_search.fit(X_train, y_train)
rc_best_accuracy = rc_grid_search.best_score_
best_parameters = rc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'criterion': 'entropy', 'max_depth': 5, 'n_estimators': 300}


In [43]:
params = [{'n_estimators':[300,310,320,330,340,350,360,370,380,390], 'criterion':['entropy'],
              'max_depth':[5,5.5,6]}]
rc_grid_search = GridSearchCV(estimator = rc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
rc_grid_search = rc_grid_search.fit(X_train, y_train)
rc_best_accuracy = rc_grid_search.best_score_
best_parameters = rc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'criterion': 'entropy', 'max_depth': 6, 'n_estimators': 390}


In [44]:
params = [{'n_estimators':[390,400,450,500], 'criterion':['entropy'],
              'max_depth':[6,6.5,7,7.5,8]}]
rc_grid_search = GridSearchCV(estimator = rc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
rc_grid_search = rc_grid_search.fit(X_train, y_train)
rc_best_accuracy = rc_grid_search.best_score_
best_parameters = rc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'criterion': 'entropy', 'max_depth': 8, 'n_estimators': 390}


In [45]:
params = [{'n_estimators':[390,391,392,393,394,395,396,397,398,399], 'criterion':['entropy'],
              'max_depth':[8,9,10,11]}]
rc_grid_search = GridSearchCV(estimator = rc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
rc_grid_search = rc_grid_search.fit(X_train, y_train)
rc_best_accuracy = rc_grid_search.best_score_
best_parameters = rc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'n_estimators': 390}


### **Prediction and Evaluation after Hyper Parammer tuning**

In [46]:
rc_y_pred = rc_grid_search.predict(X_test)
rc_cm = confusion_matrix(y_test, rc_y_pred)
print(classification_report(y_test, rc_y_pred))
acc_rc=accuracy_score(y_test, rc_y_pred)
acc_rc*100


              precision    recall  f1-score   support

           0       0.66      0.97      0.79       117
           1       0.96      0.56      0.70       133

    accuracy                           0.75       250
   macro avg       0.81      0.77      0.75       250
weighted avg       0.82      0.75      0.74       250



75.2

In [75]:
rc_roc=roc_auc_score(y_test,rc_y_pred)
rc_roc

0.7653749759012917

## ***4>> Building Decision Tree Classifier Model***

In [47]:
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(random_state = 0)
dtc.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

### **K Fold cross validation for performance of the model**

In [48]:
from sklearn.model_selection import cross_val_score
dtc_Kfold_accu = cross_val_score(estimator = dtc, X = X_train, y = y_train, cv = 10)
dtc_Kfold_accu=dtc_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(dtc_Kfold_accu*100))

Accuracy: 78.53 %


In [49]:
dtc_y_pred = dtc.predict(X_test)
dtc_y_pred

array([0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 0], dtype=int64)

### Evaluation of model using Confusion Matrix

In [50]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
cm = confusion_matrix(y_test, dtc_y_pred)
print(classification_report(y_test, dtc_y_pred))
dtc_acc=accuracy_score(y_test, dtc_y_pred)
dtc_acc

              precision    recall  f1-score   support

           0       0.68      0.75      0.72       117
           1       0.76      0.69      0.72       133

    accuracy                           0.72       250
   macro avg       0.72      0.72      0.72       250
weighted avg       0.72      0.72      0.72       250



0.72

### **Hyper Tuning of Parameters**

In [51]:
params = [{ 'criterion':['entropy','gini'],
            'max_depth':[3,4,5],'splitter':["best","random"],
           'max_features' :["auto", "sqrt", "log2"],
          'min_samples_split':[2,3,4],
          'ccp_alpha':[0.1,0.01,0.001,0.0001]}]
dtc_grid_search = GridSearchCV(estimator = dtc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
dtc_grid_search = dtc_grid_search.fit(X_train, y_train)
dtc_best_accuracy = dtc_grid_search.best_score_
best_parameters = dtc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'ccp_alpha': 0.001, 'criterion': 'gini', 'max_depth': 5, 'max_features': 'auto', 'min_samples_split': 2, 'splitter': 'best'}


In [52]:
params = [{ 'criterion':['gini'],
            'max_depth':[4.5,5,5.5,6,6.5],'splitter':["best"],
           'max_features' :["auto"],
          'min_samples_split':[1,2],
          'ccp_alpha':[0.001,0.0012,0.0013,0.0014,0.0015,0.0016]}]
dtc_grid_search = GridSearchCV(estimator = dtc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
dtc_grid_search = dtc_grid_search.fit(X_train, y_train)
dtc_best_accuracy = dtc_grid_search.best_score_
best_parameters = dtc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'ccp_alpha': 0.001, 'criterion': 'gini', 'max_depth': 6, 'max_features': 'auto', 'min_samples_split': 2, 'splitter': 'best'}


### **Prediction and Evaluation after Hyper Parammer tuning**

In [53]:
dtc_y_pred = dtc_grid_search.predict(X_test)
dtc_cm = confusion_matrix(y_test, dtc_y_pred)
print(classification_report(y_test, dtc_y_pred))
acc_dtc=accuracy_score(y_test, dtc_y_pred)
acc_dtc*100


              precision    recall  f1-score   support

           0       0.50      0.97      0.66       117
           1       0.86      0.14      0.23       133

    accuracy                           0.53       250
   macro avg       0.68      0.55      0.45       250
weighted avg       0.69      0.53      0.43       250



52.800000000000004

In [76]:
dtc_roc=roc_auc_score(y_test,dtc_y_pred)
dtc_roc

0.554848660111818

## ***5>> Building Support Vector Machine-Classifier Model***

In [54]:
from sklearn.svm import SVC
svc = SVC(random_state = 0)
svc.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

### **K Fold cross validation for performance of the model**

In [55]:
svc_Kfold_accu = cross_val_score(estimator = svc, X = X_train, y = y_train, cv = 10)
svc_Kfold_accu=svc_Kfold_accu.mean()
print("Accuracy: {:.2f} %".format(svc_Kfold_accu*100))

Accuracy: 79.33 %


In [56]:
svc_y_pred = svc.predict(X_test)
svc_y_pred

array([0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 0, 0], dtype=int64)

### Evaluation of model using Confusion Matrix

In [57]:
cm = confusion_matrix(y_test, svc_y_pred)
print(classification_report(y_test, svc_y_pred))
svc_acc=accuracy_score(y_test, svc_y_pred)
svc_acc

              precision    recall  f1-score   support

           0       0.67      0.92      0.77       117
           1       0.90      0.59      0.71       133

    accuracy                           0.75       250
   macro avg       0.78      0.76      0.74       250
weighted avg       0.79      0.75      0.74       250



0.748

### **Hyper Tuning of Parameters**

In [58]:
params = [{ 'C':[1,10,50,100,150,200],
            'kernel':['rbf','linear'],'gamma':["scale","auto"],
           'tol' :[0.001,0.0001,0.00001]}]
svc_grid_search = GridSearchCV(estimator = svc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
svc_grid_search = svc_grid_search.fit(X_train, y_train)
svc_best_accuracy = svc_grid_search.best_score_
best_parameters = svc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'C': 150, 'gamma': 'auto', 'kernel': 'rbf', 'tol': 0.001}


In [59]:
params = [{ 'C':[150,160,170,180,190],
            'kernel':['rbf'],'gamma':["auto"],
           'tol' :[0.001,0.002,0.003,0.004,0.005,0.006,0.007,0.008,0.009]}]
svc_grid_search = GridSearchCV(estimator = svc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
svc_grid_search = svc_grid_search.fit(X_train, y_train)
svc_best_accuracy = svc_grid_search.best_score_
best_parameters = svc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'C': 150, 'gamma': 'auto', 'kernel': 'rbf', 'tol': 0.002}


In [60]:
params = [{ 'C':[150,151,152,153,154,155,156,157,158,159],
            'kernel':['rbf'],'gamma':["auto"],
           'tol' :[0.002,0.0021,0.0022,0.0023,0.0024,0.0025,0.0026,0.0027,0.0028,0.0029]}]
svc_grid_search = GridSearchCV(estimator = svc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
svc_grid_search = svc_grid_search.fit(X_train, y_train)
svc_best_accuracy = svc_grid_search.best_score_
best_parameters = svc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'C': 150, 'gamma': 'auto', 'kernel': 'rbf', 'tol': 0.002}


In [61]:
params = [{ 'C':[148,149,150,150.1,150.2,150.3,150.4,150.5],
            'kernel':['rbf'],'gamma':["auto"],
           'tol' :[0.015,0.016,0.017,0.018,0.019,0.002]}]
svc_grid_search = GridSearchCV(estimator = svc,
                           param_grid = params,
                           scoring = 'accuracy',
                           cv = 10,
                           n_jobs = -1)
svc_grid_search = svc_grid_search.fit(X_train, y_train)
svc_best_accuracy = svc_grid_search.best_score_
best_parameters = svc_grid_search.best_params_
print("Best Accuracy: {:.2f} %".format(best_accuracy*100))
print("Best Parameters:", best_parameters)

Best Accuracy: 78.80 %
Best Parameters: {'C': 149, 'gamma': 'auto', 'kernel': 'rbf', 'tol': 0.002}


### Evaluation of model after tuning paramters

In [62]:
svc_y_pred = svc_grid_search.predict(X_test)
svc_cm = confusion_matrix(y_test, svc_y_pred)
print(classification_report(y_test, svc_y_pred))
acc_svc=accuracy_score(y_test, svc_y_pred)
acc_svc*100


              precision    recall  f1-score   support

           0       0.70      0.88      0.78       117
           1       0.86      0.67      0.75       133

    accuracy                           0.77       250
   macro avg       0.78      0.77      0.77       250
weighted avg       0.79      0.77      0.77       250



76.8

In [77]:
svc_roc=roc_auc_score(y_test,svc_y_pred)
svc_roc

0.7747574063363537

### 12. Now, compare the models and pick the ideal one. 

### 13. Try to Predict the target with maximum independent features. 

### ***Lets now compare the accuracies of all the models***

In [63]:
model_acc_comp=pd.DataFrame({'Models':['Multinomial_NB','Logistic Regression','Random Forest','Decision Tree',
                                       'Support Vector Machin'],
               'Accuracy Score':[acc_mnb,acc_lg,acc_rc,acc_dtc,acc_svc]})
print(model_acc_comp)

                  Models  Accuracy Score
0         Multinomial_NB           0.780
1    Logistic Regression           0.776
2          Random Forest           0.752
3          Decision Tree           0.528
4  Support Vector Machin           0.768


**As we can observe Multinomial Naives Bayes model has a better accuaracy compared to all ,comming to the independent feature we have only independent feature and one dependent feature,we have to retain both**

# Conclusion

***Basically  we have used 5 Classification models and there Best Accuracy of the models are as follows:***

1. Multinomial_NB Model:        **78.0 %**
    2. Logistic Regression  Model : **77.6 %**
    3. Random Forest Model :    **75.20 %**
    4. Decision Tree Model :    **52.80 %**
    5. Support Vector Machine : **76.80 %**

***As a result, we can say, Multinomial Naïve Bayes is the best fit to this dataset, gives model prediction performance with accuracy of 78%, logistic Regression and Random forest model are fair enough that are working very well on this Dataset. It gives approx. 77% and 75% accuracy respectively which is quite good***